### ISSUE TITLE
Rule Change: [no-extra-parens] assume unknown node types need to be parenthesised if they are parenthesised

### ISSUE THREAD

**Issue by bradzacher** [eslint/eslint#17173]
### What rule do you want to change?

no-extra-parens

### What change to do you want to make?

Generate fewer warnings

### How do you think the change should be implemented?

A new default behavior

### Example code

```ts
const x = (1 satisfies number).toFixed();
const y = (1 as number).toFixed();
```

### What does the rule currently do for this code?

[The rule currently errors on this TypeScript code](https://typescript-eslint.io/play/#ts=5.0.3&jsx=true&sourceType=module&code=PQKgBApgzgNglgOwC5gAJIJ4AdoGMBOcWSAtNPMsAgPZkAeS+AhiVk-hAlAFxgDkEfPmr4+YEMABQualxR0wAXjAAKAIxgoTJHCgAzONDAIArgFsARoICUAOiTUAYnDoQAJiusBuabKgoMJVUNJihjcyt8OwdnVw9vIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGYBfEWg&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA) because it doesn't understand that `TSAsExpression` and `TSSatisfiesExpression` have a lower precedence than `MemberExpression`.

This occurs because the precedence of "unknown" node types is automatically set to the maximum precedence:
https://github.com/eslint/eslint/blob/8bf550594fca6d29fab1a3453e701c1a457767e1/lib/rules/utils/ast-utils.js#L1444-L1445

So whenever the rule compares precedence it assumes the unknown node doesn't need parens

### What will the rule do after it's changed?

For existing JS code - no change at all.
For non-JS code (eg TS) that has new AST node types - the rule will assume that these new nodes must be wrapped in parentheses and will not report.

### Participation

- [X] I am willing to submit a pull request to implement this change.

### Additional comments

We do have an [extension rule](https://typescript-eslint.io/rules/no-extra-parens/), but some users do use the base rule (either because they choose to or accidentally).
This change would stop annoying false positives which will make users' lives better.

**Comment by sam3k** [comment]
Considering that `const z = (2).toFixed();` is perfectly okay and it passes; I think this is a good rule change.

**Comment by mdjermanovic** [comment]
> assume unknown node types need to be parenthesised if they are parenthesised

Makes sense to me 👍 

@nzakas what do you think?

**Comment by bradzacher** [comment]
My proposal for the change is essentially this:
```diff
 default:
-  return 20;
+  return node.type in eslintVisitorKeys ? 20 : -1;
```

WDYT?

**Comment by mdjermanovic** [comment]
> My proposal for the change is essentially this:
> 
> ```diff
>  default:
> -  return 20;
> +  return node.type in eslintVisitorKeys ? 20 : -1;
> ```
> 
> WDYT?

I think this will make sense for most places where `astUtils.getPrecedence` is used, but we should check each to see if the current assumption that unknown nodes have higher precedence seems better and in that case handle the `-1` return value separately. For example, in `prefer-exponentiation-operator` [here](https://github.com/eslint/eslint/blob/7709b14e18ad4e11c1119ed6575454243b8e7084/lib/rules/prefer-exponentiation-operator.js#L62) where we check if `a ** b` should be parenthesized based on its parent.

**Comment by bradzacher** [comment]
Maybe instead parameterise it so the usecases can choose?

```js
function getPrecedence(node, useLowestForUnknownNodes = true) {
  case (node.type) {
    // ...

    default:
     return !useLowestForUnknownNodes || node.type in eslintVisitorKeys ? 20 : -1;
  }
}
```

**Comment by mdjermanovic** [comment]
> Maybe instead parameterise it so the usecases can choose?
> 
> ```js
> function getPrecedence(node, useLowestForUnknownNodes = true) {
>   case (node.type) {
>     // ...
> 
>     default:
>      return !useLowestForUnknownNodes || node.type in eslintVisitorKeys ? 20 : -1;
>   }
> }
> ```

I think it's simpler and more flexible to always return `-1` for unknown nodes and let the caller handle it as a special case if needed for the use case.

**Comment by fasttime** [comment]
@nzakas I think your feedback is wanted: https://github.com/eslint/eslint/issues/17173#issuecomment-1543788058

**Comment by fasttime** [comment]
This change request has been accepted. @bradzacher would you like to submit a PR?

**Comment by bradzacher** [comment]
Sure - I'll look into it!
